frontend/src/registration/actions.ts

```diff
@@ -72,96 +75,102 @@ export const clearErrorMessage = (
     data
   });
-  setTimeout(() => {
-    try {
-      // TODO
-      if (data.domain === 'existing') {
-        throw Error('Test error');
-      }
-      dispatch({ type: Types.REGISTRATION_VALIDATION_SUCCESS, data });
-      if (nextStep) {
-        dispatch(push(nextStep));
-      }
-      dispatch({ type: UIActionTypes.NAVIGATE_NEXT_PAGE });
-    } catch (e) {
-      const error = {
-        domain: 'Domain already exists!'
-      };
+  const registrationData: any = { ...data };
+
+  V2Api.instancesOpenedxConfigValidate({
+    data: { ...registrationData }
+  }).then(() => {
+    // If validation succeeds, then save data and go to next step
+    dispatch({ type: Types.REGISTRATION_VALIDATION_SUCCESS, data });
+    if (nextStep) {
+      dispatch(push(nextStep));
+    }
+    dispatch({ type: UIActionTypes.NAVIGATE_NEXT_PAGE });
+  })
+  .catch((e: any) => {
+    e.json().then((feedback: any) => {
+      // If validation fails, return error to form through state
+      const error = { ...feedback };
+      // Loop at each error message and join them.
+      // Also convert keys from snake_case to camelCase
+      Object.keys(error).forEach(key => {
+        error[toCamelCase(key)] = error[key].join();
+      });
       dispatch({
         type: Types.REGISTRATION_VALIDATION_FAILURE,
         error
       });
-    }
-  }, 800);
+    });
+  });
 };
 
+
 export const submitRegistration = (
   data: RegistrationModel,
   nextStep?: string
 ): OcimThunkAction<void> => async dispatch => {
-  // Placeholder for form submit method
-  dispatch({
-    type: Types.REGISTRATION_SUBMIT,
-    data
-  });
+  const registrationData: any = { ...data };
+  let registrationFeedback: any = {};
 
-  setTimeout(() => {
-    try {
-      // TODO
-      if (data.domain === 'existing') {
-        throw Error('Test error');
-      }
-      dispatch({ type: Types.REGISTRATION_SUCCESS, data });
-      if (nextStep) {
-        dispatch(push(nextStep));
-      }
-      dispatch({ type: UIActionTypes.NAVIGATE_NEXT_PAGE });
-    } catch (error) {
-      dispatch({
-        type: Types.REGISTRATION_FAILURE,
-        error
+  // Check for privacy policy and replace by current date
+  if (registrationData.acceptTOS) {
+    delete registrationData.acceptTOS
+    registrationData.acceptedPrivacyPolicy = new Date();
+  } else {
+    registrationFeedback.acceptTOS = "You must accept the Terms of Service and Privacy policy."
+  }
+
+  // Check if password confirmation is correct
+  if (registrationData.password === registrationData.passwordConfirm) {
+    delete registrationData.passwordConfirm
+  } else {
+    registrationFeedback.passwordConfirm = "The password confirmation should match the password from the field above."
+  }
+
+  if (Object.entries(registrationFeedback).length === 0) {
+    V2Api.accountsCreate({
+      data: { ...registrationData }
+    }).then(() => {
+      // Perform authentication and create new instance
+      let x = dispatch(performLogin({
```

Comment:
@xitij2000 I need help here.

I created an user account using `accountsCreate` above, then when it succeeds, I call `performLogin` and make the request to return a JWT token. The thing is that I'm not sure how to link these tasks, I need the response from `performLogin` (or at least an updated root state) to be able to call the instance creation endpoint.

I see that the request and API response are correct, but I don't know how to chain these tasks:
1. Create account (already here)
2. Login (almost done, need to save tokens to state)
3. Create instance (not here yet)

Should I chain `.then(` and nest the logic? If not, what is the best way of doing this?

Reply:
Yup, I think it makes sense to chain the events. So it should go as follows:
1. Call the create account endpoint.
2. If successful call the login endpoint
3. If successful save the token by dispatching a LOGIN_SUCCESS event.

Now in the next steps you can resume the process.